File: library/analytics/src/main/java/org/cru/godtools/analytics/FirebaseAnalyticsService.java
package org.cru.godtools.analytics;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.os.Bundle;

import com.google.android.gms.common.wrappers.InstantApps;
import com.google.firebase.analytics.FirebaseAnalytics;

import org.cru.godtools.analytics.model.AnalyticsScreenEvent;
import org.cru.godtools.analytics.model.AnalyticsSystem;
import org.greenrobot.eventbus.EventBus;
import org.greenrobot.eventbus.Subscribe;
import org.greenrobot.eventbus.ThreadMode;

import java.lang.ref.Reference;
import java.lang.ref.WeakReference;

import androidx.annotation.AnyThread;
import androidx.annotation.MainThread;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import me.thekey.android.TheKey;
import me.thekey.android.eventbus.event.TheKeyEvent;

public class FirebaseAnalyticsService implements Application.ActivityLifecycleCallbacks {
    private static final String USER_PROP_APP_TYPE = "godtools_app_type";
    private static final String VALUE_APP_TYPE_INSTANT = "instant";
    private static final String VALUE_APP_TYPE_INSTALLED = "installed";

    private final FirebaseAnalytics mFirebaseAnalytics;
    private final TheKey mTheKey;

    @Nullable
    private Reference<Activity> mCurrentActivity;

    @MainThread
    private FirebaseAnalyticsService(@NonNull final Context context) {
        mFirebaseAnalytics = FirebaseAnalytics.getInstance(context);
        mTheKey = TheKey.getInstance(context);
        initFirebase(context);

        ((Application) context.getApplicationContext()).registerActivityLifecycleCallbacks(this);
        EventBus.getDefault().register(this);
    }

    @Nullable
    private static FirebaseAnalyticsService sInstance;

    @NonNull
    public static synchronized FirebaseAnalyticsService getInstance(@NonNull final Context context) {
        if (sInstance == null) {
            sInstance = new FirebaseAnalyticsService(context.getApplicationContext());
        }

        return sInstance;
    }

    // region Tracking Events

    @MainThread
    @Subscribe(threadMode = ThreadMode.MAIN_ORDERED)
    public void onAnalyticsEvent(@NonNull final AnalyticsScreenEvent event) {
        if (event.isForSystem(AnalyticsSystem.FIREBASE)) {
            handleScreenEvent(event);
        }
    }

    @AnyThread
    @Subscribe
    public void onTheKeyEvent(@NonNull final TheKeyEvent event) {
        updateUser();
    }

    // region ActivityLifecycleCallbacks

    @Override
    public void onActivityCreated(final Activity activity, final Bundle savedInstanceState) {}

    @Override
    public void onActivityStarted(final Activity activity) {}

    @Override
    public void onActivityResumed(final Activity activity) {
        mCurrentActivity = new WeakReference<>(activity);
    }

    @Override
    public void onActivityPaused(final Activity activity) {
        if (mCurrentActivity != null && mCurrentActivity.get() == activity) {
            mCurrentActivity = null;
        }
    }

    @Override
    public void onActivityStopped(final Activity activity) {}

    @Override
    public void onActivitySaveInstanceState(final Activity activity, final Bundle outState) {}

    @Override
    public void onActivityDestroyed(final Activity activity) {}

    // endregion ActivityLifecycleCallbacks

    // endregion Tracking Events

    private void initFirebase(@NonNull final Context context) {
        updateUser();
        mFirebaseAnalytics.setUserProperty(USER_PROP_APP_TYPE,
                                           InstantApps.isInstantApp(context) ? VALUE_APP_TYPE_INSTANT :
                                                   VALUE_APP_TYPE_INSTALLED);
    }

    @AnyThread
    private void updateUser() {
        mFirebaseAnalytics.setUserId(mTheKey.getDefaultSessionGuid());
    }

    @MainThread
    private void handleScreenEvent(@NonNull final AnalyticsScreenEvent event) {
        final Activity activity = mCurrentActivity != null ? mCurrentActivity.get() : null;
        if (activity != null) {
            mFirebaseAnalytics.setCurrentScreen(activity, event.getScreen(), null);
        }
    }
}

File: library/analytics/src/main/java/org/cru/godtools/analytics/firebase/FirebaseAnalyticsService.kt
package org.cru.godtools.analytics.firebase

import android.app.Application
import android.os.Bundle
import androidx.annotation.MainThread
import androidx.annotation.VisibleForTesting
import com.google.android.gms.common.wrappers.InstantApps
import com.google.firebase.analytics.FirebaseAnalytics
import java.util.Locale
import javax.inject.Inject
import javax.inject.Singleton
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import org.ccci.gto.android.common.okta.oidc.OktaUserProfileProvider
import org.ccci.gto.android.common.okta.oidc.net.response.grMasterPersonId
import org.ccci.gto.android.common.okta.oidc.net.response.ssoGuid
import org.cru.godtools.analytics.BuildConfig
import org.cru.godtools.analytics.model.AnalyticsActionEvent
import org.cru.godtools.analytics.model.AnalyticsBaseEvent
import org.cru.godtools.analytics.model.AnalyticsScreenEvent
import org.cru.godtools.analytics.model.AnalyticsSystem
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Subscribe
import org.greenrobot.eventbus.ThreadMode

/* Value constants */
private const val USER_PROP_APP_NAME = "cru_appname"
private const val VALUE_APP_NAME_GODTOOLS = "GodTools App"

private const val USER_PROP_APP_TYPE = "godtools_app_type"
private const val VALUE_APP_TYPE_INSTANT = "instant"
private const val VALUE_APP_TYPE_INSTALLED = "installed"

private const val USER_PROP_DEBUG = "debug"
private const val USER_PROP_LOGGED_IN_STATUS = "cru_loggedinstatus"
private const val USER_PROP_SSO_GUID = "cru_ssoguid"
private const val USER_PROP_GR_MASTER_PERSON_ID = "cru_grmasterpersonid"

private const val PARAM_APP_SECTION = "cru_sitesection"
private const val PARAM_APP_SUB_SECTION = "cru_sitesubsection"
private const val PARAM_CONTENT_LANGUAGE = "cru_contentlanguage"
const val PARAM_LANGUAGE_SECONDARY = "cru_contentlanguagesecondary"

@Singleton
class FirebaseAnalyticsService @VisibleForTesting internal constructor(
    app: Application,
    eventBus: EventBus,
    oktaUserProfileProvider: OktaUserProfileProvider,
    private val firebase: FirebaseAnalytics,
    coroutineScope: CoroutineScope = CoroutineScope(Dispatchers.Default)
) {
    @Inject
    @MainThread
    internal constructor(
        app: Application,
        eventBus: EventBus,
        oktaUserProfileProvider: OktaUserProfileProvider
    ) : this(app, eventBus, oktaUserProfileProvider, FirebaseAnalytics.getInstance(app))

    // region Tracking Events
    init {
        eventBus.register(this)
    }

    @MainThread
    @Subscribe(threadMode = ThreadMode.MAIN_ORDERED)
    fun onAnalyticsEvent(event: AnalyticsBaseEvent) {
        if (event.isForSystem(AnalyticsSystem.FIREBASE) || event.isForSystem(AnalyticsSystem.ADOBE)) when (event) {
            is AnalyticsScreenEvent -> handleScreenEvent(event)
            is AnalyticsActionEvent -> handleActionEvent(event)
        }
    }

    @MainThread
    private fun handleScreenEvent(event: AnalyticsScreenEvent) {
        val bundle = Bundle().apply {
            putString(FirebaseAnalytics.Param.SCREEN_NAME, event.screen)
            putString(USER_PROP_APP_NAME, VALUE_APP_NAME_GODTOOLS)
            event.locale?.let { putString(PARAM_CONTENT_LANGUAGE, it.toLanguageTag()) }
            event.appSection?.let { putString(PARAM_APP_SECTION, it) }
            event.appSubSection?.let { putString(PARAM_APP_SUB_SECTION, it) }
        }
        firebase.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW, bundle)
    }

    @MainThread
    private fun handleActionEvent(event: AnalyticsActionEvent) {
        val params = Bundle().apply {
            event.locale?.let { putString(PARAM_CONTENT_LANGUAGE, it.toLanguageTag()) }
            event.appSection?.let { putString(PARAM_APP_SECTION, it) }
            event.appSubSection?.let { putString(PARAM_APP_SUB_SECTION, it) }
            putAll(event.firebaseParams)
        }
        firebase.logEvent(event.firebaseEventName, params)
    }

    init {
        oktaUserProfileProvider.userInfoFlow(refreshIfStale = false)
            .onEach {
                firebase.setUserId(it?.ssoGuid)
                firebase.setUserProperty(USER_PROP_LOGGED_IN_STATUS, "${it != null}")
                firebase.setUserProperty(USER_PROP_GR_MASTER_PERSON_ID, it?.grMasterPersonId)
                firebase.setUserProperty(USER_PROP_SSO_GUID, it?.ssoGuid)
            }
            .launchIn(coroutineScope)

        firebase.setUserProperty(USER_PROP_APP_NAME, VALUE_APP_NAME_GODTOOLS)
        firebase.setUserProperty(PARAM_CONTENT_LANGUAGE, Locale.getDefault().toLanguageTag())
        firebase.setUserProperty(
            USER_PROP_APP_TYPE, if (InstantApps.isInstantApp(app)) VALUE_APP_TYPE_INSTANT else VALUE_APP_TYPE_INSTALLED
        )
        firebase.setUserProperty(USER_PROP_DEBUG, BuildConfig.DEBUG.toString())
    }
}
